kola/tests/update/update.go

```diff
@@ -211,3 +222,86 @@ func splitNewlineEnv(envs string) map[string]string {
 	}
 	return m
 }
+
+func sysextBootLogic(c cluster.TestCluster) {
+	// The first test case is to not use Ignition which means that the
+	// set of systemd units in the initrd is different and we also
+	// don't have Ignition mount the OEM partition
+	noIgnition, err := c.NewMachine(nil)
+	if err != nil {
+		c.Fatalf("creating test machine: %v", err)
+	}
+	version := string(c.MustSSH(noIgnition, `set -euo pipefail; grep -m 1 "^VERSION=" /usr/lib/os-release | cut -d = -f 2`))
+	if version == "" {
+		c.Fatalf("Assertion for version string failed")
+	}
+	// We disable systemd-sysext because the sysext files are empty and will fail to load.
+	// We test the following cases that differ from the test case covered by Ignition.
+	// We set up symlinks to emulate that a previous sysext was active and we store the new sysext
+	// in the rootfs instead of the OEM partition. The previous sysext is either
+	// a) stored on the rootfs and will stay there and the new one is is moved to the OEM partition
+	// b) stored on the OEM partition and gets moved to the rootfs and the new one is moved to the OEM partition
+	_ = c.MustSSH(noIgnition, fmt.Sprintf(`set -euxo pipefail
+sudo systemctl mask --now systemd-sysext
+sudo mkdir -p /etc/flatcar/sysext /etc/flatcar/oem-sysext /usr/share/oem/sysext /etc/extensions
+echo ID=test | sudo tee /usr/share/oem/oem-release
+echo myext | sudo tee /etc/flatcar/enabled-sysext.conf
+sudo touch /usr/share/oem/sysext/active-oem-test /etc/flatcar/oem-sysext/oem-test-%[1]s.raw /etc/flatcar/oem-sysext/oem-test-1.2.3.raw /etc/flatcar/sysext/flatcar-myext-%[1]s.raw
+sudo ln -fs /etc/flatcar/oem-sysext/oem-test-1.2.3.raw /etc/extensions/oem-test.raw
+`, version))
+	if err := noIgnition.Reboot(); err != nil {
+		c.Fatalf("couldn't reboot: %v", err)
+	}
+	// Check that the right symlinks are set up for case "a)" and prepare the next boot
+	_ = c.MustSSH(noIgnition, fmt.Sprintf(`set -euxo pipefail
+[ "$(readlink -f /etc/extensions/oem-test.raw)" = "/usr/share/oem/sysext/oem-test-%[1]s.raw" ] || { echo "OEM symlink wrong"; exit 1 ; }
+[ "$(readlink -f /etc/extensions/flatcar-myext.raw)" = "/etc/flatcar/sysext/flatcar-myext-%[1]s.raw" ] || { echo "Extension symlink wrong"; exit 1; }
+sudo mv /usr/share/oem/sysext/oem-test-%[1]s.raw /etc/flatcar/oem-sysext/
+sudo mv /etc/flatcar/oem-sysext/oem-test-1.2.3.raw /usr/share/oem/sysext/
+sudo ln -fs /usr/share/oem/sysext/oem-test-1.2.3.raw /etc/extensions/oem-test.raw
+`, version))
+	if err := noIgnition.Reboot(); err != nil {
+		c.Fatalf("couldn't reboot: %v", err)
+	}
+	// Check that the boot logic set up the right sysext symlinks for case "b)"
+	testCmds := fmt.Sprintf(`set -euxo pipefail
+[ "$(readlink -f /etc/extensions/oem-test.raw)" = "/usr/share/oem/sysext/oem-test-%[1]s.raw" ] || { echo "OEM symlink wrong"; exit 1 ; }
+[ "$(readlink -f /etc/extensions/flatcar-myext.raw)" = "/etc/flatcar/sysext/flatcar-myext-%[1]s.raw" ] || { echo "Extension symlink wrong"; exit 1; }
```

Comment:
Should there be a test if the old oem sysext was indeed moved to rootfs?

Reply:
Thanks!